### ISSUE TITLE
STI HAS_ONE THROUGH ISSUE

### ISSUE THREAD

**Issue by jith51** [rails/rails#22157]
Hi,
I recently posted on the issue [6663](https://github.com/rails/rails/issues/6663).
But I realized that it was closed.
So I open a new issue (which has the same content )  and join the report_bug_template below.

``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '4.2.3'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :hardwares, force: true  do |t|
    t.string :type
    t.references :host
  end
end

class Hardware < ActiveRecord::Base

end

class Computer < Hardware
  has_one :mothercard, foreign_key: :host_id
  has_one :chip, through: :mothercard
end

class Mothercard < Hardware
  belongs_to :computer, foreign_key: :host_id
  has_one :chip, foreign_key: :host_id
end

class Chip < Hardware
  belongs_to :mothercard, foreign_key: :host_id
end

class BugTest < Minitest::Test
  def test_association_chip_in_computer
    computer = Computer.create!

    mothercard = Mothercard.create(computer: computer)
    chip = Chip.create(mothercard: mothercard)

    assert_equal mothercard.id, computer.mothercard.try(:id)
    assert_equal chip.id, mothercard.chip.try(:id)
    assert_equal chip.id, computer.chip.try(:id)
  end
end
```

Below the wrong generated SQL for the last assertion : 

``` SQL
    SELECT  "hardwares".*
      FROM "hardwares"
INNER JOIN "hardwares" "mothercards_chip_join"
        ON "hardwares"."host_id" = "mothercards_chip_join"."id" 
```

``` SQL
     WHERE "hardwares"."type" IN ('Mothercard') 
```

``` SQL
       AND "hardwares"."type" IN ('Chip') 
       AND "mothercards_chip_join"."host_id" = ? 
    LIMIT 1  [["host_id", 1]]
```

And the expected SQL : 

``` SQL
    SELECT  "hardwares".*
      FROM "hardwares"
INNER JOIN "hardwares" "mothercards_chip_join"
        ON "hardwares"."host_id" = "mothercards_chip_join"."id" 
```

``` SQL
     WHERE "mothercards_chip_join"."type" IN ('Mothercard') 
```

``` SQL
       AND "hardwares"."type" IN ('Chip') 
       AND "mothercards_chip_join"."host_id" = ? 
    LIMIT 1  [["host_id", 1]]
```

**Comment by naw** [comment]
Also see

https://github.com/rails/rails/issues/12474

This is partially fixed in master, but not completely fixed in master, and definitely not fixed in 4.2

I commented on this previously here: https://github.com/rails/rails/pull/18805#issuecomment-76864493 which I will repeat here:

======
Only a subset of this issue is fixed on master.

I believe this works on master only as a side effect to changes in how relation merges work, not as an actual fix to the underlying problem.

Neither this PR (#18805) or master address the underlying problem, which is ThroughAssociations need to be smarter about pulling in default scope and STI type for the reflection chain so that it can account for the fact that the table name for the joined table might be different than its usual name due to the main association using the same table name.

Rather than skipping the condition altogether (as this PR does), the condition needs o be applied against the aliased able name assigned by the AssociationScope.

Supposing this was fixed on master, what is the procedure to backport to 4-2-stable?

Here is a gist using the ActiveRecord bug report template with failing tests for 4.2 and failing tests for master:

https://gist.github.com/naw/fa55816dae7805355efd
